Correct DQL `INSTANCE OF` to filter all possible child classes

Jean85 opened the pull request:
This is the rebased version of #1441, which fixes #4646.
Here is the original text:

> There was a bug in the "INSTANCE OF" operator as described in
> https://groups.google.com/forum/#!topic/doctrine-user/B8raq8CNMgg
> 
> "INSTANCE OF" was not taking into account subclasses.
> It was merely translating the class to its discriminator.
> This is not correct since the class can have subtypes and those
> are, indeed, still instance of the superclass.
> Also, classes may not have a discriminator (e.g. abstract classes).
> 
> This commit also provides useful tests to avoid regression.

taueres committed:
[QUERY] "INSTANCE OF" now behaves correctly with subclasses

There was a bug in the "INSTANCE OF" operator as described in
https://groups.google.com/forum/#!topic/doctrine-user/B8raq8CNMgg

"INSTANCE OF" was not taking into account subclasses.
It was merely translating the class to its discriminator.
This is not correct since the class can have subtypes and those
are, indeed, still instance of the superclass.
Also, classes may not have a discriminator (e.g. abstract classes).

This commit also provides useful tests to avoid regression.

taueres committed:
[QUERY] "INSTANCE OF" now behaves correctly with subclasses

There was a bug in the "INSTANCE OF" operator as described in
https://groups.google.com/forum/#!topic/doctrine-user/B8raq8CNMgg

"INSTANCE OF" was not taking into account subclasses.
It was merely translating the class to its discriminator.
This is not correct since the class can have subtypes and those
are, indeed, still instance of the superclass.
Also, classes may not have a discriminator (e.g. abstract classes).

This commit also provides useful tests to avoid regression.

phansys commented on tests/Doctrine/Tests/ORM/Functional/InstanceOfTest.php:
```diff
@@ -0,0 +1,119 @@
+        {
+            parent::setUp();
+
+            $this->_schemaTool->createSchema(array(
```
Please, use short array syntax instead.

phansys commented on tests/Doctrine/Tests/ORM/Functional/InstanceOfAbstractTest.php:
```diff
@@ -0,0 +1,113 @@
+            $this->assertCount(1, $result);
+
+            foreach ($result as $r) {
+                $this->assertInstanceOf('Doctrine\Tests\ORM\Functional\InstanceOfAbstractTest\Person', $r);
```
Please, use `::class` constant instead of hardcoding the QCN here.

phansys commented on tests/Doctrine/Tests/ORM/Functional/InstanceOfAbstractTest.php:
```diff
@@ -0,0 +1,113 @@
+            foreach ($result as $r) {
+                $this->assertInstanceOf('Doctrine\Tests\ORM\Functional\InstanceOfAbstractTest\Person', $r);
+                $this->assertInstanceOf('Doctrine\Tests\ORM\Functional\InstanceOfAbstractTest\Employee', $r);
+                $this->assertEquals('bar', $r->getName());
```
Please, use `assertSame()` instead.

phansys reviewed (changes requested):
Status: Needs work.

Jean85 committed:
Fix as per review

Jean85 committed:
Fix as per review

Jean85 commented:
All done, thanks for the review @phansys

lcobucci reviewed (changes requested):
Can you please extract some methods from `SqlWalker#walkInstanceOfExpression()` it was already messy and now it's a bit harder to understand things properly.

Also could you try to relate the tests, and entities that each test use, to the ticket (as we have on `Functional\Ticket` classes) just to keep the "standard" we're having...

Jean85 committed:
Fix small CS issues as per review

Jean85 committed:
Fix small CS issues as per review

Jean85 committed:
Split SqlWalker::walkInstanceOfExpression method

Jean85 committed:
Split SqlWalker::walkInstanceOfExpression method

Jean85 committed:
Move tests to ticket namespace (and rename them)

Jean85 committed:
Move tests to ticket namespace (and rename them)

Jean85 commented:
Thanks for the review @lcobucci!
I hope that I addressed al your concerns with my latest commits.

Jean85 committed:
Fix test

Jean85 committed:
Fix test

Jean85 commented:
A lot of tests are failing, probably for an unrelated issue... should I rebase? Or merge master?

Ocramius commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
+                $currentMetadata = $this->em->getClassMetadata($class);
+                $currentDiscriminator = $currentMetadata->discriminatorValue;
+
+                if (is_string($currentDiscriminator) && ! array_key_exists($currentDiscriminator, $discriminators)) {
```
A discriminator may be also an integer or similar - why do you want this to only work with `string`?

Jean85 replied:
I didn't know! I'll do a `!== null` then.

Ocramius commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
+            }
+
+            // Trim first backslash
+            $parameter = ltrim($parameter, '\\');
```
You should do `$this->em->getClassMetadata($parameter);` here - this already provides the `ltrim()` logic for you (internally)

Jean85 replied:
Right! I was also already doing the `getClassMetadata` a few lines after, I'll just move that up!

Ocramius commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
+            // Trim first backslash
+            $parameter = ltrim($parameter, '\\');
+
+            if ($parameter !== $discrClass->name && ! array_key_exists($parameter, $knownSubclasses)) {
```
`$metadata->getClass()->isSubclassOf(...)` maybe?

Ocramius replied:
This also has the effect of getting rid of `INSTANCE OF` checks where `INSTANCE OF` is pointing at the topmost class (which is most likely a bug in userland code)

Ocramius commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
+                $currentDiscriminator = $currentMetadata->discriminatorValue;
+
+                if (is_string($currentDiscriminator) && ! array_key_exists($currentDiscriminator, $discriminators)) {
+                    $discriminators[$currentDiscriminator] = true;
```
Consider using `null` if that `true` has no meaning

Ocramius commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
+            $sqlParameterList[] = $this->conn->quote($dis);
+        }
+
+        return '(' . implode(', ', $sqlParameterList) . ')';
```
According to this logic, the list may be empty, leading to a crash

Jean85 replied:
How that could happen? Since [I check if the class is unrelated at line 2297](https://github.com/doctrine/doctrine2/pull/6392/files#diff-0c8825d1265a66bb8b20f3d99c276961R2297), at least one discriminator should pop up, or an exception be thrown...

Ocramius replied:
Yeah, but that flow makes it hard to follow: ideally we should `array_map()` over `$instanceOfExpr->value`, but let's see where the code goes after the additional tests provided by @taueres

Ocramius commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
+        }
+
+        foreach (array_keys($discriminators) as $dis) {
+            $sqlParameterList[] = $this->conn->quote($dis);
```
`array_map([$this->conn, 'quote'], array_keys($discriminators))`

Jean85 replied:
We may already have values inside `$sqlParameterList`, see [line 2289](https://github.com/doctrine/doctrine2/pull/6392/files#diff-0c8825d1265a66bb8b20f3d99c276961R2289), so I think it's better to do it like that.

Ocramius replied:
Then just `array_merge()` with it

Ocramius commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php:
```diff
@@ -0,0 +1,114 @@
+<?php
+
+namespace Doctrine\Tests\ORM\Functional\Ticket {
```
Please put the entire file under a single namespace

Ocramius commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php:
```diff
@@ -0,0 +1,156 @@
+         */
+        private $specialization;
+
+        public function getSpecialization()
```
Remove all accessors that aren't strictly needed: prefer public properties if they are needed (although note that these tests don't really need the properties to work, so you should drop everything but the identifier)

Ocramius commented on tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php:
```diff
@@ -502,7 +502,7 @@ public function testSupportsInstanceOfExpressionsInWherePart()
         $this->assertSqlGeneration(
             "SELECT u FROM Doctrine\Tests\Models\Company\CompanyPerson u WHERE u INSTANCE OF Doctrine\Tests\Models\Company\CompanyEmployee",
-            "SELECT c0_.id AS id_0, c0_.name AS name_1, c0_.discr AS discr_2 FROM company_persons c0_ WHERE c0_.discr IN ('employee')"
+            "SELECT c0_.id AS id_0, c0_.name AS name_1, c0_.discr AS discr_2 FROM company_persons c0_ WHERE c0_.discr IN ('manager', 'employee')"
```
Ouch, this is a bad bug. Good catch!

Jean85 replied:
All credits to @taueres !

Jean85 committed:
Refactor a bit the SqlWalker modifications

Jean85 committed:
Refactor a bit the SqlWalker modifications

Jean85 committed:
Apply additional fixes to the SqlWalker to fix tests

Jean85 committed:
Apply additional fixes to the SqlWalker to fix tests

Jean85 committed:
Put all tests classes in a single namespace

Jean85 committed:
Put all tests classes in a single namespace

Jean85 committed:
Simplify stubs used in tests

Jean85 committed:
Simplify stubs used in tests

Jean85 commented:
Ok seems that I've addressed all your concerns, @Ocramius @lcobucci 
Build is green with low deps, other failures seems unrelated, maybe they come from master?

taueres commented:
Hi! Thank you very much for resuming this PR, I really appreciate that!
I checked this fix once more but unfortunately the issue still occurs when using `INSTANCE OF` with parameters.

Here is an example:
```php
$dql = 'SELECT p FROM Person p WHERE p INSTANCE OF :parameter';
$query = $em->createQuery($dql);
$query->setParameter('parameter', $em->getClassMetadata(Person::class));

// $result contains only direct instances of `Person`, no subclasses
$result = $query->getResult();
```

taueres commented:
I wrote a failing test case for the issue I mentioned before: https://gist.github.com/taueres/05ff035bb2f72cddd7711b7f93cd9d6a

Jean85 commented:
Thanks @taueres!
I've found that your test case fails also when doing this:

```php
$dql = 'SELECT p FROM Doctrine\Tests\ORM\Functional\Ticket\PersonTicket4646Parametric p
        WHERE p INSTANCE OF :parameter';
$query = $this->_em->createQuery($dql);
$query->setParameter(
    'parameter',
    PersonTicket4646Parametric::class
);
```
Do we want to check for this additional uses, @Ocramius ?

Ocramius commented:
```php
$query->setParameter(
    'parameter',
    PersonTicket4646Parametric::class
);
```

This currently requires binding a `ClassMetadata` instance:

```php
$query->setParameter(
    'parameter',
    $em->getClassMetadata(PersonTicket4646Parametric::class)
);
```

We can fix it, yes

Jean85 committed:
Adding a failing test case for parameter binding using metadata with INSTANCE OF

Jean85 committed:
Adding a failing test case for parameter binding using metadata with INSTANCE OF

Jean85 commented:
Ok I've added the failing test case. I'll try to work on that soon; @taueres if you have any tips, I'm all ears! :smile:

taueres commented:
Unfortunately solving the parameter issue is not as easy as static DQL queries.
The main problem is that **Doctrine is always translating each DQL parameter to one single SQL parameter**.
For `INSTANCE OF`, this 1-to-1 relationship is not always true: one class might be mapped to multiple discriminators.

The issue is not easy to address because the parameter values could still be undefined while Doctrine is parsing the query.

Think about the following scenario:
```php
$dql = 'SELECT p FROM Person p WHERE p INSTANCE OF :parameter';
$sql = $query->getSQL();
// $sql is something like "SELECT * FROM person WHERE person.kind IN (?)"
// which is WRONG because we don't know yet how many SQL parameters are needed.
```

Do you have any idea of how we could tackle this problem?
Is DBAL supporting `IN` operator with array parameters?

Ocramius commented:
> Is DBAL supporting `IN operator with array parameters?

Yep, but it only supports [these two types of bindings](https://github.com/doctrine/dbal/blob/03e17189c48a92bff9c5fd7d9242d13a15780d15/lib/Doctrine/DBAL/Connection.php#L73-L85), so custom discriminators may be a problem

Commit:
Fix discriminator resolution when using parameters

taueres commented:
Okay, I have a solution! `INSTANCE OF` with parameters now works correctly on my branch https://github.com/taueres/doctrine2/tree/fix-instance-of-subclasses

It still needs some minor improvements, e.g. duplicated methods.

How should I proceed with this PR? Should I create a new one from my branch?

Ocramius commented:
@taueres I suggest waiting for @Jean85 to merge the changes into his branch

Jean85 committed:
Merge additional fix (and master changes) from taueres/fix-instance-of-subclasses

Jean85 commented:
I've merged your fixes @taueres, thanks a lot!
Now the only issue that I see is the duplication of the `getAllDiscriminators(ClassMetadata $classMetadata)` method in both the `Query` and `SqlWalker` classes. Any suggestion on how to address that? It uses `$this->getEntityManager()->getClassMetadata($class)`, which I think is the biggest hurdle to overcome, otherwise it could be moved inside the `ClassMetadata` class.

taueres commented:
To solve the duplication issue we could create something like:
```php
class HierarchyDiscriminatorResolver
{
    public static function resolveDiscriminatorsForClass(ClassMetadata $rootClassMetadata, EntityManagerInterface $entityManager)
    {
        /* Duplicate code goes here */
        /* return list of discriminator */
    }
}
```

In the proposed solution, the `EntityManager` dependency is needed in order to retrieve `ClassMetadata` for subClasses.

I would not put such logic into `ClassMetadata` because atm `ClassMetadata` is not coupled to `EntityManager`.

Jean85 commented:
This seems a feasible solution for me... @Ocramius ? What do you think?

Ocramius commented:
@Jean85 as long as the class is marked as `@internal`, sure :+1: 

Totally missed these comments before btw.

Jean85 committed:
Remove code duplication of the getAllDiscriminators method

Jean85 commented:
Don't worry @Ocramius, I was on vacation anyway :smile: 
Deduplication of code done, build is green, this is ready for a final review!

Ocramius commented on lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php:
```diff
@@ -0,0 +1,34 @@
+ * @package Doctrine\ORM\Utility
+ * @internal This class exists only to avoid code duplication, do not reuse it externally
+ */
+class HierarchyDiscriminatorResolver
```
This class needs to be tested in isolation

Ocramius replied:
`final`

Ocramius commented on lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php:
```diff
@@ -0,0 +1,34 @@
+use Doctrine\ORM\EntityManagerInterface;
+
+/**
+ * Class HierarchyDiscriminatorResolver
```
Remove this line

Ocramius commented on lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php:
```diff
@@ -0,0 +1,34 @@
+
+/**
+ * Class HierarchyDiscriminatorResolver
+ * @package Doctrine\ORM\Utility
```
Remove this line

Ocramius commented on lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php:
```diff
@@ -0,0 +1,34 @@
+ */
+class HierarchyDiscriminatorResolver
+{
+    public static function resolveDiscriminatorsForClass(
```
Add a `private` constructor

Ocramius replied:
An explanation of the method is needed

Ocramius commented on lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php:
```diff
@@ -0,0 +1,34 @@
+    public static function resolveDiscriminatorsForClass(
+        ClassMetadata $rootClassMetadata,
+        EntityManagerInterface $entityManager
+    ) {
```
Return type

Jean85 replied:
Sorry, I forgot that the target branch is 7+! Good!

Ocramius commented on lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php:
```diff
@@ -0,0 +1,34 @@
+        $hierarchyClasses[] = $rootClassMetadata->name;
+
+        $discriminators = [];
+        foreach ($hierarchyClasses as $class) {
```
Empty line before logical blocks

Jean85 replied:
You should enforce this in the Lint stage of the CI!

lcobucci replied:
We're going to :wink

lcobucci commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -2297,4 +2269,37 @@ public function walkResultVariable($resultVariable)
+     * @return string The list in parentheses of valid child discriminators from the given class
+     * @throws QueryException
+     */
+    private function getChildDiscriminatorsFromClassMetadata(ClassMetadataInfo $rootClass, AST\InstanceOfExpression $instanceOfExpr)
```
Please add the  return type declaration

taueres commented:
According to the DQL spec. Users can provide multiple classes for an INSTANCE OF statement.
```
InstanceOfExpression     ::= IdentificationVariable ["NOT"] "INSTANCE" ["OF"] (InstanceOfParameter | "(" InstanceOfParameter {"," InstanceOfParameter}* ")")
```
Do we have some tests covering this specific use case?

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php:
```diff
@@ -0,0 +1,64 @@
+
+class Ticket4646InstanceOfAbstractTest extends OrmFunctionalTestCase
+{
+    protected function setUp()
```
`: void`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php:
```diff
@@ -0,0 +1,64 @@
+        ]);
+    }
+
+    public function testInstanceOf()
```
`: void`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php:
```diff
@@ -0,0 +1,64 @@
+        $query = $this->_em->createQuery($dql);
+        $result = $query->getResult();
+
+        $this->assertCount(1, $result);
```
Please use `self::assert*` for new tests

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php:
```diff
@@ -0,0 +1,77 @@
+
+class Ticket4646InstanceOfMultiLevelTest extends OrmFunctionalTestCase
+{
+    protected function setUp()
```
`: void`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php:
```diff
@@ -0,0 +1,77 @@
+        ]);
+    }
+
+    public function testInstanceOf()
```
`: void`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php:
```diff
@@ -0,0 +1,77 @@
+        $query = $this->_em->createQuery($dql);
+        $result = $query->getResult();
+
+        $this->assertCount(3, $result);
```
Here too (`self::assert*`)

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php:
```diff
@@ -0,0 +1,67 @@
+
+class Ticket4646InstanceOfParametricTest extends OrmFunctionalTestCase
+{
+    protected function setUp()
```
`: void`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php:
```diff
@@ -0,0 +1,67 @@
+        ]);
+    }
+
+    public function testInstanceOf()
```
`: void`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php:
```diff
@@ -0,0 +1,67 @@
+            $this->_em->getClassMetadata(PersonTicket4646Parametric::class)
+        );
+        $result = $query->getResult();
+        $this->assertCount(2, $result);
```
`self::assert*` here too

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php:
```diff
@@ -0,0 +1,66 @@
+
+class Ticket4646InstanceOfTest extends OrmFunctionalTestCase
+{
+    protected function setUp()
```
`: void`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php:
```diff
@@ -0,0 +1,66 @@
+        ]);
+    }
+
+    public function testInstanceOf()
```
`: void`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php:
```diff
@@ -0,0 +1,77 @@
+     */
+    private $id;
+
+    public function getId()
```
`: ?int`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php:
```diff
@@ -0,0 +1,67 @@
+     */
+    private $id;
+
+    public function getId()
```
`: ?int`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php:
```diff
@@ -0,0 +1,66 @@
+        $query = $this->_em->createQuery($dql);
+        $result = $query->getResult();
+
+        $this->assertCount(2, $result);
```
Here too

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php:
```diff
@@ -0,0 +1,66 @@
+     */
+    private $id;
+
+    public function getId()
```
`: ?int`

lcobucci reviewed (dismissed):
Some nitpicking 😄

Jean85 committed:
Apply various and CS fixes as per review

Jean85 committed:
Add unit test for HierarchyDiscriminatorResolverTest

Jean85 committed:
Add more CS fixes

Jean85 commented:
@lcobucci @Ocramius done as per review!

Jean85 committed:
Add new functional test to check usage of INSTANCEOF with multiple parameters

Jean85 commented:
I've also added a new functional test to cover the case of multiple parameters passed to the `INSTANCEOF`, as suggested by @taueres

Ocramius reviewed (approved):
LGTM :+1: :ship:

Ocramius commented:
Thanks @Jean85 @taueres and @phansys!
